### PR TITLE
replace catastrophically slow jq with a Python script

### PR DESCRIPTION
**Pull request by infinitewarp**
This is necessary because `jq` `1.6` is **very** slow (10x or worse by my measurements) compared to `jq` `1.5`. Its unexpected slowness is causing pod liveness and readiness probes to exceed timeouts, and that results in frequent pod restarting. Apparently the performance regression was addressed in `jq` master (https://github.com/stedolan/jq/issues/1826#issuecomment-464480110, https://github.com/stedolan/jq/pull/1834) *more than three years ago*, but there has been no release since then, and there appears to be no current release plan (https://github.com/stedolan/jq/issues/2305). The future of the `jq` project is unclear; so, we should find something else that works and that we can rely on in the future. 😕 

I'm proposing we stop using `jq` in our probes and instead use this small Python script to extract Clowder configs into the environment once, write a more quickly processed file, and reuse that file in subsequent calls.

## Performance comparison

I very roughly benchmarked the performance differences by locally running Docker with a `cdappconf.json` copied from an ephemeral environment deployment.

### current postigrade image using jq 1.6 (slow)
```
❯ docker run --privileged -v $(pwd)/cdapp:/cdapp -e ACG_CONFIG=/cdapp/cdappconfig.json -it --entrypoint sh quay.io/cloudservices/postigrade:ab29854

sh-4.4# jq --version
jq-1.6
sh-4.4# env | grep -c ^DB_
0
sh-4.4# time for i in {1..100}; do source /clowder_init.sh > /dev/null; done

real	0m25.530s
user	0m24.403s
sys	0m2.205s
sh-4.4# env | grep -c ^DB_
7
```

### old postigrade image using jq 1.5 (fast)
```
❯ docker run --privileged -v $(pwd)/cdapp:/cdapp -e ACG_CONFIG=/cdapp/cdappconfig.json -it --entrypoint sh quay.io/cloudservices/postigrade:bf02247

sh-4.4# jq --version
jq-1.5
sh-4.4# env | grep -c ^DB_
0
sh-4.4# time for i in {1..100}; do source /clowder_init.sh > /dev/null; done && echo

real	0m3.019s
user	0m2.258s
sys	0m1.745s
sh-4.4# env | grep -c ^DB_
7
```

### with Python script and new tmp file (very fast)
```
❯ docker build -q -t postigrade:latest .
sha256:7458367c439b30b402645ef52423b493ea1511c11eb5735e763c3a40d8217974
❯ docker run --privileged -v $(pwd)/cdapp:/cdapp -e ACG_CONFIG=/cdapp/cdappconfig.json -it --entrypoint sh postigrade:latest

sh-4.4# env | grep -c ^DB_
0
sh-4.4# rm -f /tmp/clowder_envs.sh
sh-4.4# time for i in {1..100}; do source /clowder_init.sh > /dev/null; done

real	0m0.150s
user	0m0.124s
sys	0m0.024s
sh-4.4# env | grep -c ^DB_
7
```

### with Python script, but destroy tmp file each time

```
❯ docker build -q -t postigrade:latest .
sha256:7458367c439b30b402645ef52423b493ea1511c11eb5735e763c3a40d8217974
❯ docker run --privileged -v $(pwd)/cdapp:/cdapp -e ACG_CONFIG=/cdapp/cdappconfig.json -it --entrypoint sh postigrade:latest

sh-4.4# env | grep -c ^DB_
0
sh-4.4# time for i in {1..100}; do rm -f /tmp/clowder_envs.sh; source /clowder_init.sh > /dev/null; done

real	0m8.556s
user	0m7.394s
sys	0m1.020s
sh-4.4# env | grep -c ^DB_
7
```

The *first* time `/clowder_init.sh` is called, the Python script runs and is still about 3x slower than our old approach using jq 1.5, but it's also about 3x faster than the current jq 1.6. *Subsequent* calls to `/clowder_init.sh` are *very* fast because they're just reading the file that the Python script wrote in the first iteration.

Average results over 100 iterations:
- current image with jq 1.6: **255 ms** per `/clowder_init.sh` call
- older image with jq 1.5: **30 ms** per `/clowder_init.sh` call
- proposed image in this PR: **2 ms** per `/clowder_init.sh` call

How do the probes perform in an actual ephemeral environment deployment?

```
sh-4.4$ time /probe-liveness.sh 
localhost:8000 - accepting connections

real    0m0.008s
user    0m0.004s
sys     0m0.006s
sh-4.4$ time /probe-readiness.sh 
 psql_is_ready 

real    0m0.012s
user    0m0.004s
sys     0m0.007s
```

See also our recent discussions in Slack:
- https://ansible.slack.com/archives/C8VGAPJNN/p1655929184616189?thread_ts=1655924376.591149&cid=C8VGAPJNN
- https://ansible.slack.com/archives/C8VGAPJNN/p1655995461647799
- https://coreos.slack.com/archives/C022YV4E0NA/p1655930541439289?thread_ts=1655476360.723759&cid=C022YV4E0NA

### CODE REVIEW

**Comment by abellotti** [comment]
At first I wasn't crazy about the extra script and temp file, I was thinking more of something like this to mimic the jq usage:

```

function pjq() {
    python -c "
import sys, json
from flatten_dict import flatten
json_data = flatten(json.loads(sys.stdin.read()), reducer='dot', enumerate_types=(list,))
key = \"${1}\".lstrip('.')
value = json_data[key] if key in json_data.keys() else \"\"
print(value)
"
}

ACG_CONFIG="cdappconfig.json"

export DB_NAME="`cat $ACG_CONFIG | pjq '.database.name'`"
export DB_HOST="`cat $ACG_CONFIG | pjq '.database.hostname'`"
export DB_PORT="`cat $ACG_CONFIG | pjq '.database.port'`"
export DB_USER="`cat $ACG_CONFIG | pjq '.database.username'`"
export DB_PASSWORD="`cat $ACG_CONFIG | pjq '.database.password'`"
export DB_SSLMODE="`cat $ACG_CONFIG | pjq '.database.sslMode'`"
export DB_CERT="`cat $ACG_CONFIG | pjq '.database.rdsCa'`"
export PG_BOUNCER_LISTEN_PORT="`cat $ACG_CONFIG | pjq '.webPort'`"
```

However, invoking python several times wasn't too performant even with fs caching. So we'll go with your approach.

The only thing is if you could test with a stage cdappconfig.json if you haven't already just to make sure rdsCa is properly fetched. I will send you a copy a have from earlier ssl work.

**Comment by infinitewarp** [comment]
> At first I wasn't crazy about the extra script and temp file

For what it's worth, I feel the same way. Writing the tmp file was the very last idea and addition I made as I was noodling around here. Invoking the Python script is generally *much* slower than invoking jq, and the only reason it keeps pace with our previous images is that I'm only calling the Python script *once* whereas we were calling jq *eight* times in each probe. 😖 

I was wary about writing to the running container's filesystem at this point in the process, but it *does* seem to work consistently both locally and deployed in OpenShift, and it should recover just fine if the file is unexpectedly deleted between runs. I think I'm comfortable enough with it now.

> The only thing is if you could test with a stage cdappconfig.json if you haven't already just to make sure rdsCa is properly fetched. I will send you a copy I have from earlier ssl work.

Thanks for sharing that, @abellotti! 👍  I ran my new code with that file locally, and I think the results look good, but I also DM'd you a privatebin link with details to review and verify yourself.

File: json_to_env.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python3
+"""
+Read JSON file and dump contents to source-able shell script.
+
+Useful for extracting Clowder's $ACG_CONFIG file (usually '/cdapp/cdappconf.json').
```
Comment (abellotti) [inline json_to_env.py:5]: minor nit-pick, it's actually /cdapp/cdappconfig.json

**Comment by abellotti** [comment]
@infinitewarp should we bring back the original timing with this PR ?

**Comment by infinitewarp** [comment]
> @infinitewarp should we bring back the original timeouts with this PR ?

I considered about this earlier too, and I was thinking about changing them in another PR later. One step at a time just to be safe, especially since this is already a big change.

File: json_to_env.py (hunk 1 above)
Comment (infinitewarp) [reply json_to_env.py:5]: Good catch! Everywhere else just uses the env var, but I should fix this docstring to be as accurate as possible.

**Review by abellotti — DISMISSED** [review]

**Pull request merged**